src/core/builder.rs

```diff
@@ -0,0 +1,354 @@
+use std::collections::HashSet;
+use std::mem;
+use std::path::Path;
+use std::fs::File;
+use std::io::prelude::*;
+use log::{debug, info, warn};
+
+use pyo3::prelude::*;
+
+use crate::importlib::*;
+use crate::minimal_parser::*;
+use super::types::*;
+
+pub struct BuildMetadata {
+    pub processed: usize,
+    pub from_cache: usize,
+}
+
+impl BuildMetadata {
+    pub fn new() -> BuildMetadata {
+        BuildMetadata {
+            processed: 0,
+            from_cache: 0,
+        }
+    }
+}
+
+#[pyclass]
+pub struct GraphBuilder {
+    pub graph: DepGraph,
+    processing: HashSet<String>,
+    verbose: bool,
+    cache: Option<DepGraph>,
+    metadata: BuildMetadata
+}
+
+#[pymethods]
+impl GraphBuilder {
+    #[new]
+    pub fn new(verbose: Option<bool>) -> Self {
+        let verbose = if let Some(verbose) = verbose {
+            verbose
+        } else {
+            false
+        };
+
+        let builder = GraphBuilder {
+            graph: DepGraph::new(),
+            processing: HashSet::new(),
+            verbose: verbose,
+            cache: None,
+            metadata: BuildMetadata::new(),
+        };
+
+        builder
+    }
+
+    pub fn build(&mut self, source: &str, package: Option<String>) -> DepGraph {
+        // Trying to make source look like a package
+        let (package, dirs) = if let Some(package) = package {
+            (package, Some(vec![]))
+        } else {
+            ("<terminal>".to_string(), None)
+        };
+
+        // Manually build spec / DepNode for first call
+        let name = "<terminal>".to_string();
+        let spec = ModuleSpec {
+            name: name.clone(),
+            origin: None,
+            // Treating this as the main file which is not a package
+            parent: package,
+            submodule_search_locations: dirs
+        };
+
+
+        let node = DepNode::new(spec.clone(), Some(0));
+        self.graph.add(node);
+        self._process_imports(spec, source);
+
+        if self.metadata.from_cache == 0 {
+            info!(
+                "Processed {} dependency relationships.",
+                self.metadata.processed
+            );
+        } else {
+            info!(
+                "Processed {} dependency relationships ({} from cache).",
+                self.metadata.processed,
+                self.metadata.from_cache
+            );
+        }
+
+        // Reset for next build
+        let _ = mem::replace(&mut self.processing, HashSet::new());
+        let _ = mem::replace(&mut self.metadata, BuildMetadata::new());
+        let graph = mem::replace(&mut self.graph, DepGraph::new());
+
+        // Cache all nodes
+        let to_cache = graph.clone();
+        if let Some(cache) = self.cache.as_mut() {
+            cache.merge(to_cache)
+        } else {
+            self.cache = Some(to_cache)
+        }
+
+        return graph
+    }
+}
+
+impl GraphBuilder {
+    pub fn _load_source(&mut self, node: &DepNode) -> Option<String> {
```

Comment:
I mentioned this last weekend, but if the `_` prefix is meant to denote that the function is private, you can remove the `pub` prefix on the function definition to enforce the function is private